### DISCUSSION TITLE
Question about "Object of type Row is not JSON serializable" exception triggered on 1.4.31

### DISCUSSION THREAD

**Discussion by evilaliv3** [sqlalchemy/sqlalchemy#7939] [Usage Questions]
Hello,

i've a question in relation to the question "Object of type Row is not JSON serializable".

I'm testing the compatibility of @globaleaks with SQLAlchemy 1.4.3 that currently works on SQLAlchemy 1.3

Apparently that when running on SQLAlchemy 1.4.3 some queries are deferred so that my objects get a reference to a Row object instead of the actual result.

Do you have any advice on the possible change in SQLAlchemy that could have caused this?

Thank you for all your great work!

**Comment by zzzeek** [comment]
assuming you are using future=True, Row has been changed to resemble a NamedTuple much more closely: https://docs.sqlalchemy.org/en/14/changelog/migration_14.html#rowproxy-is-no-longer-a-proxy-is-now-called-row-and-behaves-like-an-enhanced-named-tuple

if you want dictionaries, use either `result.mappings()` to get a `MappedResult` that will return dictionary `MappedRow` objects, or each `Row` also has a `Row._mapping` attribute that does the same.

**Comment by evilaliv3** [comment]
Thank you so much for your advice.

What do you recommend to implement code that is compatible with both 1.3 and 1.4?

As for your questions i actually do not know if i'm using future=True or what it is but i will check; I'm currently using sqlalchemy as installed on Ubuntu 22.02 to which i'm posting GlobaLeaks.

**Reply by zzzeek** [reply]
you can use future=False until you can upgrade, or you can convert rows using a function like

```py
from sqlalchemy import __version__
if __version__.startswith("1.3"):
    def cvt(row):
        return dict(row)
else:
   def cvt(row):
        return row._mapping
```

**Comment by evilaliv3** [comment]
Thank you so much, this is very helpful.

Actually i will need to have the full codebase compatible with both 1.3 and 1.4.3 for long term as we install sqlalchemy as a debian package so that the version depend on the distribution.

Im thinking to a possible monkeypatch of Sqlalchemy to add .mapping() to the Result object on 1.3 this way i can just write all in line with 1.4.3 reducing the number of if/else.

would this work in your opinion?

**Comment by evilaliv3** [comment]
I've actually found a better solution:

```
try:
    from sqlalchemy.engine import row
except:
    row = None

class JSONEncoder(json.JSONEncoder):
    def default(self, obj):
        if row && isinstance(obj, row.Row):
            return dict(obj)
        else:
            return super().default(obj)
```

so that if i get a Row (that means i'm on sqlalchemy 1.4.3) i can convert it just where i could not use it.

Thank you for your support @zzzeek

**Comment by CaselIT** [comment]
Just as a note, I would advise against using version 1.4.3, but instead use a more recent one since early 1.4.x had quite a few issues that were fixed in later versions.

Any reason for using that specific version?

**Comment by evilaliv3** [comment]
Thank you @CaselIT , i always talked about 1.4.3 but actually talking about 1.4.31 that is the version included in the upcoming Ubuntu 22.04

**Reply by CaselIT** [reply]
understood!
